src/query/phrase_query/phrase_weight.rs

```diff
@@ -72,13 +73,53 @@ impl PhraseWeight {
                 }
             }
         }
-        Ok(Some(PhraseScorer::new(
+        Ok(Some(Box::new(ExactPhraseScorer::new(
             term_postings_list,
             similarity_weight,
             fieldnorm_reader,
             self.scoring_enabled,
             self.slop,
-        )))
+        ))))
+    }
+
+    fn phrase_scorer_as_scorer(
+        &self,
+        reader: &SegmentReader,
+        boost: Score,
+    ) -> crate::Result<Option<Box<dyn Scorer>>> {
+        let similarity_weight = self.similarity_weight.boost_by(boost);
+        let fieldnorm_reader = self.fieldnorm_reader(reader)?;
+        let mut term_postings_list = Vec::new();
+        if reader.has_deletes() {
+            for &(offset, ref term) in &self.phrase_terms {
+                if let Some(postings) = reader
+                    .inverted_index(term.field())?
+                    .read_postings(term, IndexRecordOption::WithFreqsAndPositions)?
+                {
+                    term_postings_list.push((offset, postings));
+                } else {
+                    return Ok(None);
+                }
+            }
+        } else {
+            for &(offset, ref term) in &self.phrase_terms {
+                if let Some(postings) = reader
+                    .inverted_index(term.field())?
+                    .read_postings_no_deletes(term, IndexRecordOption::WithFreqsAndPositions)?
+                {
+                    term_postings_list.push((offset, postings));
+                } else {
+                    return Ok(None);
+                }
+            }
+        }
+        Ok(Some(Box::new(ExactPhraseScorer::new(
```

Comment:
Obviously here you'd have if self.slop > 0 -> ExactPhraseScorer, else SlopPhraseScorer